### PR TITLE
Add a 2v2 arena variant of Hanabi

### PR DESCRIPTION
**Pull request by lipovetz**

### CODE REVIEW

File: kaggle_environments/envs/open_spiel_env/games/hanabi/harness.py [hunk 1]
```diff
@@ -458,48 +481,173 @@ def _format_move_history(lines: Sequence[str] | None) -> str:
...
 )
 
+# A slot written as a negative index -- "Play -1". _normalize drops the hyphen
+# (which is what makes "Discard slot-2" work), so without this guard the move
+# silently resolves to slot 1. A model writing -1 most plausibly means a
+# Python-style index from the END of the hand, which is the opposite end from
+# slot 1 -- the engine would take a move the model did not choose and no
+# rethink would fire. Anchored at the verb so only the operand position counts:
+# "Discard slot-2" and a trailing "Play 0 - 2 lives left" are untouched.
+_NEGATIVE_SLOT_RE = re.compile(r"^\W*(?:play|discard)\s+-\s*\d", re.IGNORECASE)
```
Comment (kaggle-agent) [inline kaggle_environments/envs/open_spiel_env/games/hanabi/harness.py]: `_NEGATIVE_SLOT_RE = r"^\W*(?:play|discard)\s+-\s*\d"` requires the sign to sit *immediately* after the verb, but `_RAW_SLOT_RE` (line 444) explicitly allows a run of `_FILLER` words (`slot|card|my|the|from|in|number`) between verb and operand. So any filler defeats the guard and the move lands on the wrong card — the precise failure the comment above says it exists to prevent.

Verified by execution against `legal = ["(Play 0)","(Play 1)","(Discard 0)","(Discard 1)"]`:

```
'Play -1'         -> None          # guard fires, correct
'Play slot -1'    -> '(Play 1)'    # guard bypassed, silent wrong move
'Play the -1'     -> '(Play 1)'
'Discard slot -1' -> '(Discard 1)'
```

`_normalize` drops the hyphen, `_RAW_SLOT_RE` absorbs `slot`/`the` as filler, and slot 1 is submitted. A model writing `-1` most plausibly means the far end of the hand, so this is the opposite card, with no rethink and no telemetry.

Fix: let the guard walk the same filler run the slot regex does, e.g.

```python
_NEGATIVE_SLOT_RE = re.compile(
    rf"^\W*(?:play|discard)(?:\s+{_FILLER})*\s+-\s*\d", re.IGNORECASE
)
```

(`_FILLER` is defined above `_NEGATIVE_SLOT_RE` in both files, so the f-string reference is fine.) Worth a case per filler word in the parser tests.

**Also affects** `kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py:705` — the matcher block is byte-identical there; please fix both.

File: kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py [hunk 2]
```diff
@@ -0,0 +1,1092 @@
...
+        seat = int(state.get("your_seat", player_id % players_per_team))
+    except (TypeError, ValueError):
+        return {}
+    if players_per_team < 2:
+        return {}
+    team_base = team_id * players_per_team
+    return {team_base + (seat + offset) % players_per_team: offset for offset in range(1, players_per_team)}
+
+
+def _resolve_offset(
```
Comment (kaggle-agent) [inline kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py:892]: The bare-number fallback in `_resolve_offset` is sound for base Hanabi, where every player id at the table is also a plausible offset reading, but arena ids span **two** tables and the `as_offset = number if offsets == {number} else None` branch doesn't know that.

Verified from seat P3 (`player_id_offsets == {2: 1}`, `legal = [..., "(Reveal player +1 rank 3)"]`):

```
'Reveal player 2 rank 3' -> '(Reveal player +1 rank 3)'  # correct: P2 is the teammate
'Reveal player 1 rank 3' -> '(Reveal player +1 rank 3)'  # P1 is on the OTHER table
'Reveal player 0 rank 3' -> None                          # correctly refused
```

`as_player_id` is correctly `None` for id 1 (not in the offset map), but `as_offset` rescues it because the only legal offset happens to be `1`. A model that names an opposing-table seat — the one naming mistake this game newly makes possible, since the other table is invisible — gets a hint fired at its teammate instead of a rethink. Note the reading is also genuinely ambiguous for the model: with `players_per_team == 2` the sole legal offset is always `+1`, so the number `1` is simultaneously a valid relative offset and an out-of-table absolute id.

Suggest suppressing the offset reading whenever the observation was available and the number is a known-but-wrong player id, i.e. when `player_id_offsets` is non-empty and `number` falls in `range(players_per_team * num_teams)` but isn't a key. Keep the current behaviour when `player_id_offsets` is `None` (no observation), which `test_the_observation_is_optional` already pins.

**Review by kaggle-agent — COMMENTED** [review]
**review-harness**: The diff is mostly catalogue-sanctioned hardening (closed filler list, second-action guard, annotation splitting, hint-slot walk-forward) and the arena prompt's rules all check out against hanabi_arena_game.py / hanabi_proxy.py. One real gameplay bug: the new negative-slot guard is anchored so tightly to the verb that any filler word defeats it, so "Play slot -1" still silently resolves to slot 1 — the exact wrong-move the guard was added to stop (verified by execution in both hanabi/harness.py and hanabi_arena/harness.py). Secondary: the arena's bare-number offset fallback can resolve an opposing-table player id to a hint against the teammate. Rest are accuracy/dead-code suggestions.

File: kaggle_environments/envs/open_spiel_env/games/hanabi/harness.py [hunk 3]
```diff
@@ -458,48 +481,173 @@ def _format_move_history(lines: Sequence[str] | None) -> str:
...
+# those is the same false rejection the second-action guard above was narrowed
+# to avoid.
+#
+# "instead OF" is excluded even at the front, because the preposition inverts
+# the word: bare "instead" offers an alternative ("Play 0 -- instead, a hint"),
+# while "instead of X" names the alternative being REJECTED ("Play 0 -- instead
+# of burning a token"). The second is a settled answer with its rationale
+# attached, which is the most natural way to write one down.
+_COLOR_VALUE = r"(?:red|yellow|green|white|blue|colou?rs?\b|[rygwb]\b)"
 _UNDECIDED_RE = re.compile(
```
Comment (kaggle-agent) [inline kaggle_environments/envs/open_spiel_env/games/hanabi/harness.py:617]: `_ANNOTATION_RE` is `re.DOTALL` but `_UNDECIDED_RE`'s first branch gates on `^(?=.*(?:\d|\bor\b))`, where `.` is *not* newline-safe. A second action split across a line break therefore reads as harmless commentary:

```
'Play 0, 1'    -> None          # refused, correct
'Play 0,\n1'   -> '(Play 0)'    # accepted, same content
```

Same input, different verdict purely on whitespace — and a numbered/bulleted list of candidate moves (the common shape for a model that hasn't committed) is exactly the multi-line case. Adding `re.DOTALL` to `_UNDECIDED_RE`, or `(?s)` on that lookahead, makes the two agree. Worth confirming the existing "annotation is legitimately multi-line" tests still pass, since this makes the undecided check strictly stricter.

**Also affects** `kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py:691`.

File: kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py [hunk 4]
```diff
@@ -0,0 +1,1092 @@
...
+    # only for that seat -- the legal hints against a hand are exactly the
+    # colors and ranks IN it, so another seat's move list would spell out cards
+    # the reader must not see.
+    #
+    # There is deliberately no serialized-state fallback below this. Every
+    # other OpenSpiel harness deserializes `serializedGameAndState` when the
+    # observation is thin, but `hanabi_arena` withholds that blob from agents
+    # precisely because it reconstructs every hidden hand; reaching for it here
+    # would be asking for the thing the env refused to send. If neither source
+    # above has a move list, {} costs the turn -- which is the same cost the
```
Comment (kaggle-agent) [inline kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py]: The comment says "If neither source above has a move list, `{}` costs the turn -- which is the same cost the fallback carried when the blob was unreadable." That isn't what happens. In `core_harness.create_agent_fn`, an empty `get_legal_moves` result with `playerId`/`currentPlayer` populated (always true for an acting arena seat) fires `_TELEMETRY(no_legal_actions=True)` and then `raise ValueError("No legal actions available.")` — the agent errors out rather than losing a turn, and in a team game that propagates to a team-scoped forfeit via `team_of()`.

The decision not to reach for `serializedGameAndState` is right and well argued; only the stated consequence is wrong. Please reword so the next reader doesn't treat `{}` as a soft failure.

File: kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py [hunk 5]
```diff
@@ -0,0 +1,1092 @@
...
+    given = entry.get("slots_when_given")
+    if given is None:
+        # Which slots a hint pointed at is the substance of the hint. A guess
+        # is worse than an omission, and "no slots" would be a lie -- the
+        # engine only allows hints that match at least one card.
+        return ""
+    current = entry.get("slots", given)
+    noun = "slot" if len(given) == 1 else "slots"
+    if not given:
+        return " -- no slots"
```
Comment (kaggle-agent) [inline kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py]: `return " -- no slots"` directly contradicts the comment six lines above it, which says "'no slots' would be a lie -- the engine only allows hints that match at least one card." With the current env it's unreachable (`_public_facts` never writes an empty `slots_when_given`), so this is dead code that encodes the falsehood the function elsewhere refuses to print. Suggest `return ""` for consistency with the `given is None` branch. Minor: `noun` is computed on the preceding line but unused on this path.

File: kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py [hunk 6]
```diff
@@ -0,0 +1,1092 @@
...
+
+def _describe_decision(entry: Mapping[str, Any], team_base: int) -> str:
+    """One history line: the action plus the public facts it revealed.
+
+    Every fact read here was recorded by the env at the moment the action
+    resolved, when it became common knowledge at this table (see
+    ``_public_facts`` in ``hanabi_arena_game``). Nothing still face-down is
+    read, and the opposing table never appears in this list at all.
+    """
+    player_id = entry.get("player_id", team_base)
```
Comment (kaggle-agent) [inline kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py]: `entry.get("player_id", team_base)` (and `entry.get("target_player_id", team_base)` at line 416) default a missing id to the first seat on the team, so a dropped field renders as a confident, wrong attribution — "P2 played slot 1" when the harness doesn't actually know who acted. Every neighbouring renderer in this file deliberately degrades the other way (`_readable`, `_render_discards`, `_format_move_history` all emit "unavailable"). `_public_facts` always writes both keys today, so this is unreachable, but it's the one place in the file that would lie rather than admit a gap. Suggest defaulting to `None` and falling back to the same "unavailable" wording.

File: kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py [hunk 7]
```diff
@@ -0,0 +1,1092 @@
...
+            f"  slot {i}: {_card_text(c.get('card'))} -- told {_told_text(c)}; possible {_possible_text(c)}"
+            for i, c in enumerate(cards)
+        ]
+        blocks.append(header + "\n" + "\n".join(lines))
+    return "\n".join(blocks)
+
+
+def _render_fireworks(table: Mapping[str, Any]) -> str:
+    fireworks = table.get("fireworks")
+    num_ranks = int(table.get("ranks", 5) or 5)
```
Comment (kaggle-agent) [inline kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py]: `int(table.get("ranks", 5) or 5)` hardcodes the rank count, while line 1012 in this same file reads the identical key as `int(table.get("ranks", _DEFAULT_RANKS) or _DEFAULT_RANKS)`. The key name is right (`hanabi_proxy.state_dict` emits `ranks`), so this is only a divergence risk if `_DEFAULT_RANKS` is ever retuned for a variant — the fireworks block would then silently disagree with the rest of the prompt. Use `_DEFAULT_RANKS` in both places.

File: kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py [hunk 8]
```diff
@@ -0,0 +1,1092 @@
+"""LLM harness for Hanabi Arena (2v2 team variant of OpenSpiel Hanabi).
```
Comment (kaggle-agent) [inline kaggle_environments/envs/open_spiel_env/games/hanabi_arena/harness.py:1]: Not a defect, a coverage note: this diff is a substantial regex rewrite of the answer matcher, duplicated verbatim across two harnesses. I had no replay archive available, so I verified by direct execution against hand-built corpora rather than real model output — I'm deliberately not quoting any hit rates. Before this ships, a scan of existing Hanabi replays through the new `_match_move_to_legal` (diffing old vs new resolution per response) would be the cheapest way to catch a regression in the cases my corpora didn't imagine, especially around `_split_annotation`'s `for _ in range(2)` bound and `_PREDICTION_RE` disarming genuine second actions.

**Review by c-h-i-a-m-a-k-a — APPROVED** [review]
